3rdParty/fuerte/include/fuerte/waitgroup.h

```diff
@@ -61,7 +61,7 @@ class WaitGroup {
     std::unique_lock<std::mutex> lock(_mutex);
     _counter--;
     if (_counter == 0) {
-      lock.unlock();
+      //lock.unlock();
```

Comment:
IMHO this unlock should be perfectly safe.

Reply:
I thought so too but without it I can fairly consistently produce a hanging connection on macOS

Reply:
Interesting! Let me take a closer look at this, because at the moment I really cannot see how this unlock could cause a hang.

Reply:
It really seemed like undefined behavior, but I even managed to reproduce this on two different computers.
The strangest thing is that attaching a debugger would always immediately wake up the erroneously waiting thread.

Reply:
IMHO there is nothing wrong with the unlock. All accesses to`_counter` happen under the lock, providing a total order of the operations. So either the update in `done` happens before the read in `wait` or the other way round. In the first case, `wait` will see the new value of `_counters` and return. In the second case, the call to wait _atomically_ releases the lock and blocks. But this implies that the call to `notify_all` is ordered after the call to `wait`, so the notification cannot be missed.

Did you see this issues on systems other then MacOS too? Unfortunately, attaching a debugger to a process can have various side-effects. Were you able to reproduce this when you start the process under a debugger?

Reply:
Only on MacOS. One thread would eventually get stuck on `_conditionVar.wait(...)`,
when running the process in a debugger it also happened. Doing anything in the debugger like stopping / continuing or enabling breakpoints would also magically wakup the thread.

Reply:
I can confirm that it didn't clear the situation, the typical 2 candidates are hanging again on teh mac.

Reply:
ok, it seems to be working now.